pkg/pillar/base/execwrapper.go

```diff
@@ -0,0 +1,150 @@
+	var timer <-chan time.Time
+	agentName := getAgentName()
+
+	if log != nil {
+		log.Debugf("ExecCommand(%s, %t, %t, %s, %v)", agentName, combined, timeOut, command, arg)
+	}
+
+	cmd := exec.Command(command, arg...)
+
+	var buf bytes.Buffer
+	cmd.Stdout = &buf
+	if combined {
+		cmd.Stderr = &buf
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("ExecCommand(%s, %v): Error while starting command: %s",
+			command, arg, err.Error())
+	}
+
+	// Use a channel to signal completion so we can use a select statement
+	done := make(chan error)
+	go func() { done <- cmd.Wait() }()
+	stillRunning := time.NewTicker(25 * time.Second)
+
+	if timeOut {
+		timer = time.After(timeOutLimit * time.Second)
+	} else {
+		timer = time.After(maxTimeOutLimit * time.Second)
+	}
+
+	for {
+		select {
+		case <-timer:
+			// Timeout happened first, kill the process.
+			cmd.Process.Kill()
+			return nil, fmt.Errorf("ExecCommand(%s, %v): Command timed out", command, arg)
+		case err := <-done:
+			// Command completed before timeout.
+			return buf.Bytes(), err
+		case <-stillRunning.C:
+		}
+		updateAgentTouchFile(log, agentName)
+	}
+}
+
+func updateAgentTouchFile(log *LogObject, agentName string) {
+	if agentName == "" {
+		if log != nil {
+			log.Warnf("updateAgentTouchFile: agentName is empty")
+		}
+		return
+	}
+	filename := fmt.Sprintf("/var/run/%s.touch", agentName)
+	_, err := os.Stat(filename)
+	if err != nil {
+		file, err := os.Create(filename)
+		if err != nil {
+			if log != nil {
+				log.Infof("updateAgentTouchFile: %s\n", err)
+			}
+			return
+		}
+		file.Close()
+	}
+	_, err = os.Stat(filename)
+	if err != nil {
+		if log != nil {
+			log.Errorf("updateAgentTouchFile: %s\n", err)
+		}
+		return
+	}
+	now := time.Now()
+	err = os.Chtimes(filename, now, now)
+	if err != nil {
+		if log != nil {
+			log.Errorf("updateAgentTouchFile: %s\n", err)
+		}
+		return
+	}
+}
+
+func getAgentName() string {
+	stackTrace := strings.Split(string(debug.Stack()), "\n")
+	var methodName, file, agent, resultAgentName string
+	for i, trace := range stackTrace {
+		if strings.HasPrefix(trace, "github.com/lf-edge/eve/pkg/pillar/cmd/") {
+			trace = strings.TrimPrefix(trace, "github.com/lf-edge/eve/pkg/pillar/cmd/")
+			start := strings.Index(trace, ".")
+			end := strings.Index(trace, "(")
+			if start > -1 && end > -1 {
+				methodName = strings.TrimSpace(trace[start+1 : end])
+				agent = strings.TrimSpace(trace[:start])
+			}
+			filePath := strings.TrimSpace(stackTrace[i+1])
+			start = strings.LastIndex(filePath, "/")
+			end = strings.Index(filePath, ":")
+			if start > -1 && end > -1 {
+				file = strings.TrimSpace(filePath[start+1 : end])
+			}
+			break
+		}
+	}
+	switch agent {
+	case "volumemgr", "wstunnelclient", "baseosmgr", "verifier", "downloader", "zedrouter", "zedmanager", "tpmmgr",
+		"vaultmgr", "logmanager", "nim", "ledmanager", "nodeagent", "domainmgr":
+		resultAgentName = agent
+	case "zedagent":
+		switch file {
```

Comment:
This seems like a smaller maintenance pain; means that any additional touch file in some other agent needs to be known by this code.

But if this is a short-term thing until we can factor out the Exec calls from the main go routine I think we can live with it.